### ISSUE TITLE
When Attention Meets Fast Recurrence: Training Language Models with Reduced Compute

### ISSUE THREAD

**Issue by shnakazawa** [TeamAidemy/ds-paper-summaries#14]
Lei, Tao. 2021. “When Attention Meets Fast Recurrence: Training Language Models with Reduced Compute.” In Proceedings of the 2021 Conference on Empirical Methods in Natural Language Processing, 7633–48. Online and Punta Cana, Dominican Republic: Association for Computational Linguistics. https://aclanthology.org/2021.emnlp-main.602/

- 低計算コストで言語モデリングが可能なSRU++を提案
- Attentionと高速リカレントネットワークに着目
- Transformer系の3倍から25倍少ない計算量で、同等の性能を示した

Introductionで **"Is attention all we need for modeling?"** と、あの有名な言葉をもじり、実際にAttentionを挟むだけで大幅な改善を見せ、その威力を見せてくれる面白い仕事。

## Abstract

>Large language models have become increasingly difficult to train because of the growing computation time and cost. In this work, we present SRU++, a highly-efficient architecture that combines fast recurrence and attention for sequence modeling. SRU++ exhibits strong modeling capacity and training efficiency. On standard language modeling tasks such as Enwik8, Wiki-103 and Billion Word datasets, our model obtains better bits-per-character and perplexity while using 3x-10x less training cost compared to top-performing Transformer models. For instance, our model achieves a state-of-the-art result on the Enwik8 dataset using 1.6 days of training on an 8-GPU machine. We further demonstrate that SRU++ requires minimal attention for near state-of-the-art performance. Our results suggest jointly leveraging fast recurrence with little attention as a promising direction for accelerating model training and inference.

大規模な言語モデルは、計算時間やコストが増大するため、学習が困難になってきている。本研究では、シーケンスモデリングのために高速な再帰性と注意力を組み合わせた高効率なアーキテクチャであるSRU++を発表する。SRU++は、強力なモデリング能力と学習効率を発揮する。Enwik8、Wiki-103、Billion Wordデータセットなどの標準的な言語モデリングタスクにおいて、我々のモデルは、上位のTransformerモデルと比較して、3倍から10倍少ない学習コストで、優れた文字あたりのビット数とパープレキシティを獲得することができる。例えば、Enwik8データセットでは、8GPUのマシンで1.6日間の学習を行い、最先端の結果を達成しました。さらに、SRU++は最小限の注意で最新鋭に近い性能を発揮することを実証しています。この結果は、モデルの学習と推論を高速化するための有望な方向性として、少ない注意で高速リカレンスを共同で活用することを示唆しています。

## コード

https://github.com/asappresearch/sru

## 解決した課題/先行研究との比較

- 近年の言語モデルはデータセットのサイズ、アーキテクチャのサイズともに大規模化してきており、計算コストがどんどん高くなってきている。
- 近年の研究からAttentionと高速リカレントネットワークを組み合わせることで、より効率的なモデルを構築できる可能性が示唆されていた (Bradbury et al., 2017; Zhang and Sennrich, 2019)
- 本論文ではそのアイデアを検証した。
	- 2018年に報告した[SRU](https://arxiv.org/abs/1709.02755) (Lei et al., 2018) をベースにしている。

## 技術・手法のポイント

[![Image from Gyazo](https://i.gyazo.com/96ed23604cb1fa034305cae81ec835c9.png)](https://gyazo.com/96ed23604cb1fa034305cae81ec835c9)

- リカレントネットワーク (SRU) の入力の線形変換を、Attentionを用いた変換に置換
- 具体的には、ベースとなったSRUでは $\mathbf{U}$ を以下のように計算している。

$$
\displaystyle\mathbf{U}^{\top}\ \displaystyle=\ \left(\begin{array}[]{l}\mathbf{W}\\
\mathbf{W}^{\prime}\\
\mathbf{W}^{\prime\prime}\end{array}\right)\mathbf{X}^{\top}
$$

- 一方で、SRU++ では内部でAttentionを用いて以下の計算を行っている。

$$
\begin{array}{l}
\mathbf{Q} \displaystyle=\mathbf{W}^{q}\mathbf{X}^{\top} \\
\mathbf{K} \displaystyle=\mathbf{W}^{k}\mathbf{Q}^{\top} \\
\mathbf{V} \displaystyle=\mathbf{W}^{v}\mathbf{Q}^{\top}
\end{array}
$$

$$
\displaystyle\mathbf{A}^{\top}=\text{softmax}\left(\frac{\mathbf{Q}^{\top}\mathbf{K}}{\sqrt{d^{\prime}}}\right)\mathbf{V}^{\top}.
$$

$$
\displaystyle\mathbf{U}^{\top}=\mathbf{W}^{o}\left(\mathbf{Q}+\alpha\cdot\mathbf{A}\right).
$$

- 線形変換からより複雑性の高い演算に変更することで表現力が向上
- 後段のリカレントネットワーク箇所はSRUと同様
- これが何故効果的にはたらくの？の議論は **Fig.5** 周辺を参照。

## 評価指標

- ENWIK8 (**Table 3**), WIKI103 (**Table 5**), BILLION WORD (**Table 6**) データセットを用い、Transformer系のモデルと比較。
- 評価指標はPerplexityとGPU days. (ENWIK8はBPC)
- すべてのデータセットにおいて、Transformer系の3倍から25倍少ない計算量で、同等の結果を達成。
- また、WIKI-103データセットの推論スピードが数倍速い。 (**Table 7**) 
- IWSLT'14のDe→En翻訳タスクでもTransfomrerと同程度の性能。 (**Table 8**) 
- Attentionを挿入する場所についても検討したところ、入力側の1層目が最も効果的であった。 (**Fig.4**)

## 残された課題・議論

- Attention自体の高速化とは異なるアイデアであるため、Attention自体の高速化と合わせれば、より全体の高速化が進められるかもしれない。
- 同様に、本提案手法はリカレント実装や正規化、最適化技術の進歩と組み合わせられると考えられる。
- Introductionではモデルのサイズにも触れられていたが、パラメータ数が多いことに変わりはなし。パラメータ数がネックになる場合にはこの手法は使えない。(**Table 3**)
	- 「軽量化されたためエッジデバイスにも載せやすい！」という方向を目指した仕事ではない。

## 重要な引用
- Transformer
	- Vaswani, Ashish, Noam Shazeer, Niki Parmar, Jakob Uszkoreit, Llion Jones, Aidan N. Gomez, Lukasz Kaiser, and Illia Polosukhin. 2017. “Attention Is All You Need.” arXiv [cs.CL]. arXiv. http://arxiv.org/abs/1706.03762.
- Attentionと高速リカレントネットワークを組み合わせた効率的なモデル構築の可能性を示唆
	-  Bradbury, James, Stephen Merity, Caiming Xiong, and Richard Socher. 2016. “Quasi-Recurrent Neural Networks.” arXiv [cs.NE]. arXiv. http://arxiv.org/abs/1611.01576.
	- Zhang, Biao, and Rico Sennrich. 2019. “A Lightweight Recurrent Network for Sequence Modeling.” In Proceedings of the 57th Annual Meeting of the Association for Computational Linguistics, 1538–48. Florence, Italy: Association for Computational Linguistics.
- Tramsformer-XL. 比較対象となっている、論文投稿時点での最強モデル
	- Dai, Zihang, Zhilin Yang, Yiming Yang, Jaime Carbonell, Quoc V. Le, and Ruslan Salakhutdinov. 2019. “Transformer-XL: Attentive Language Models Beyond a Fixed-Length Context.” arXiv [cs.LG]. arXiv. http://arxiv.org/abs/1901.02860.
- SRU. 同著者の先行研究
	- Lei, Tao, Yu Zhang, Sida I. Wang, Hui Dai, and Yoav Artzi. 2017. “Simple Recurrent Units for Highly Parallelizable Recurrence.” arXiv [cs.CL]. arXiv. http://arxiv.org/abs/1709.02755.

## 参考情報

- [Transformer日本語要約](https://github.com/TeamAidemy/ds-paper-summaries/issues/1)
